### PR TITLE
Improving error msg logic

### LINKED ISSUE #749: unclear logic when handling errors
it is unclear what the logic should be in `error_message.py`.

there is a `_is_syntax_error` function that's supposedly capturing syntax errors, but it also captures things like "catalog error", "not found", and "does not exist", which doesn't seem syntax errors, but rather seem like errors displayed when the user tries to query a non-existent table (which is different than a SQL syntax error).

Since `_is_syntax_error` is used to determine whether to show the "If using snippets, you may pass the --with argument..." error message suggestion, it seems that the purpose of the function is to capture "missing table errors" as opposed to "syntax errors". In such a scenario, we shouldn't be capturing syntax errors.

I modified `test_sql_error_suggests_using_cte` in https://github.com/ploomber/jupysql/pull/746 to reflect this. some tests are marked as xfail because I'm unsure what the purpose of this code is.

### PR DESCRIPTION
**Pull request by neelasha23**
## Describe your changes

Improving error handling logic.

## Issue number

Closes #749 

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [ ] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [ ] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--782.org.readthedocs.build/en/782/

### CODE REVIEW

**Comment by neelasha23** [comment]
Please review @edublancas

File: src/sql/magic.py [hunk 1]
```diff
@@ -577,11 +576,21 @@ def interactive_execute_wrapper(**kwargs):
             StatementError,
         ) as e:
             # Sqlite apparently return all errors as OperationalError :/
-            self._error_handling(e, command.sql)
+            detailed_msg = detail(e)
+            self._error_handling(e, command.sql, detailed_msg)
```
Comment (edublancas) [inline src/sql/magic.py:580]: this code looks pretty convoluted. do we need recursion here? `_error_handling` is calling `_error_handling`.

**Review by edublancas — CHANGES_REQUESTED** [review]
the functions have more descriptive names but the overall code is still convoluted. it's hard to follow since there are recursive calls

File: src/sql/magic.py (hunk 1 above)
Comment (neelasha23) [reply src/sql/magic.py:580]: There isn't recursion in [_error_handling](https://github.com/ploomber/jupysql/blob/8a9582a6f2ef9055479d76621dc0063104f28764/src/sql/magic.py#L235).

Comment (neelasha23) [reply src/sql/magic.py:580]: I'm moving all this code to a `ErrorHandler` class. Can I close this PR and merge with the circular dependency PR since I would need that here also? @edublancas

Comment (edublancas) [reply src/sql/magic.py:580]: oh yeah. I forgot that we had to do this because of the circular dependency. yes, I'll close it now
